### ISSUE TITLE
Unexpected default behavior of QueueClient.OnMessage (ReceiveAndDelete instead of PeekLock)

### ISSUE THREAD

**Issue by ncarandini** [Azure/azure-service-bus-dotnet#97]
## Preamble
`ClientQueue` has a property `Mode` of type `ReceiveMode`:

```csharp
public enum ReceiveMode
 {
    // Specifies the PeekLock receive mode.
    // This is the default value for Microsoft.ServiceBus.Messaging.ReceiveMode.
    PeekLock = 0,
    // Specifies the ReceiveAndDelete receive mode.
    ReceiveAndDelete = 1
}
```
and the `PeekLock` mode is indeed the default mode when requesting a message from the client, i.e. with something like this (from the [Apps that use queues](https://docs.microsoft.com/en-us/azure/service-bus-messaging/service-bus-create-queues) example):

```csharp
MessageReceiver receiver = factory.CreateMessageReceiver("DataCollectionQueue");
BrokeredMessage receivedMessage = receiver.Receive();
try
{
    ProcessMessage(receivedMessage);
    receivedMessage.Complete();
}
catch (Exception e)
{
    receivedMessage.Abandon();
}
```

## Issue
When using `QueueClient.OnMessage(Action<BrokeredMessage> callback)` to subscribe to the queue, things work differently than expected, and the `ClientQueue.Mode` property is ignored and the default behavior is the `ReceiveAndDelete`.

This is undocumented, as far as I have seen, and the only [example of `OnMessage` usage in the documentation](https://docs.microsoft.com/en-us/azure/service-bus-messaging/service-bus-dotnet-get-started-with-queues) simply omit the Complete() method invocation without explaining why.

The only way to have the `PeekLock` behavior with `OnMessage` is to explicitly declare `AutoComplete = false` in the `OnMessageOptions`:

```csharp
Client.OnMessage((receivedMessage) =>
{
    try
    {
        // Process the message
        receivedMessage.Complete();
    }
    catch (Exception e)
    {
        // Handle any message processing specific exceptions here
        receivedMessage.Abbandon();
    }
}, new OnMessageOptions() { AutoComplete = false });
```

## Conclusions
I understand that this is not a bug but an unexpected default behavior.
Is unexpected because:
- everywhere in the documentation is stated that the default behavior is the PeekLock mode.
- the OnMessage ignore the value of the  `ClientQueue.Mode` property in favor of the `OnMessageOptions.AutoComplete` property.

## Suggested action
I also understand that changing the default `OnMessageOptions.AutoComplete` value to `false` could produce breaking changes on code actually in production, so my only request is to stress this issue in the documentation clearly stating that the default behavior of `OnMessage` is the `ReceiveAndDelete` mode .

## UPDATE:
Please see the comment following the reopened action, that is related to this repo, where I ask for a change of the default behavior and I propose the code changes, based on the fact that this is a new library, still in preview.

**Comment by SeanFeldman** [comment]
@ncarandini 👍 for better documentation. The current client (not this repo) is doing a [horrible job](https://docs.microsoft.com/en-us/dotnet/api/microsoft.servicebus.messaging.onmessageoptions#Microsoft_ServiceBus_Messaging_OnMessageOptions_AutoComplete).

But to be fair, this client in this repo is providing the [xmldoco](https://github.com/Azure/azure-service-bus-dotnet/blob/0c8c472102633f078518808e3185ac1745e60efa/src/Microsoft.Azure.ServiceBus/OnMessageOptions.cs#L21), which is an improvement 😃 

This is why I had #26 from day one.
/cc @jtaubensee @nemakam

**Comment by ncarandini** [comment]
Ops, I totally missed it!
I just found the [How can we improve the Windows Azure Service Bus?](https://feedback.azure.com/forums/216926-service-bus/suggestions/6948939-open-source-microsoft-servicebus-dll-windowsazure) where you explain that the current client is not on GitHub.

Thanks a lot for your clarifying comment.

I'll close this issue as is unrelated to this repo. Sorry for the mess.

**Comment by ncarandini** [comment]
Hi @jtaubensee, I cloned this repo and saw that the same issue is present here.
Now that I've more things to say, pointing to specific code lines of `MessageReceiver.cs` , `OnMessageOptions.cs`, `MessageReceivePump.cs` and `QueueClient.cs`.
Do you prefer me to open a new issue or reopen this one, removing (by you, if you'll think is ok) of the invalid tag ?

**Comment by ncarandini** [comment]
Thanks @jtaubensee,
I'll procede here :-)

**Comment by jtaubensee** [comment]
Hi @ncarandini - 

So the default behavior is still peek lock, however we default the value `OnMessageSettings.AutoComplete` to true. This means that when the `OnMessage` callback exits, `Message.Complete` is called. If you were to call `Message.Abandon` or `Message.Defer` inside of your `OnMessage` callback, the message would still exist on the queue.

`ReceiveDelete` means that when the message is retrieved from the queue, it is also deleted from the queue. While this performance can be better, it can result in message loss, if your receiver dies while processing the message.

Does that make sense? To your point, our documentation could do a better job of explaining this.

**Comment by ncarandini** [comment]
As stated before, I do think that the default mode of the OnMessage should be PeekLook.
Here are my proposals:

### Set AutoComplete to false as the default value
Looking at the `OnMessageOptions.cs` we see that the constructor set the default options:

```csharp
public OnMessageOptions()
{
    this.MaxConcurrentCalls = 1;
    this.AutoComplete = true;
    this.ReceiveTimeOut = Constants.DefaultOperationTimeout;
    this.AutoRenewTimeout = Constants.ClientPumpRenewLockTimeout;
}
```
The easiest things will be to change it to:

```csharp
...
this.AutoComplete = false;
...
```
or to add a line in the `Microsoft.Azure.ServiceBusCostants` class and reference it:
```csharp
...
this.AutoComplete = Constants.DefaultAutoComplete;
...
```

```csharp
namespace Microsoft.Azure.ServiceBus
{
    using System;
    internal static class Constants
    {
        ...
        public const bool DefaultAutoComplete = false;
        ...
    }
}
```
### Inherit the value  of AutoComplete from the Mode value of the `MessageReceiver` (of the `QueueClient`)

From now on, I'll refer only to the `MessageReceiver`, because calling the `QueueClient.OnMessageAsync()` method calls the same method of the `QueueClient.InnerReceiver` (of type  `MessageReceiver`) and because I take for granted that the `QueueClient.InnerReceiver.Mode` property will have the same value of the `QueueClient.Mode` one.

That being said, calling te `MessageReceiver.OnMessageAsync()` method without the optional `OnMessageOptions` parameter, will actually produce a call to the more complete method passing a new instance of `OnMessageOptions` class with the default values created as we've seen before with the notable exception of the `ReceiveTimeOut`, that is inherited from the value of the corresponding property of the caller (i.e. the `MessageReceiver` instance):

```csharp
public void OnMessageAsync(Func<BrokeredMessage, CancellationToken, Task> callback)
{
    this.OnMessageAsync(callback, new OnMessageOptions() {
        ReceiveTimeOut = this.OperationTimeout
    });
}
```

My suggestion is to inherit also the `AutoComplete` from the caller `Mode` value, with a very simple type conversion:

```csharp
public void OnMessageAsync(Func<BrokeredMessage, CancellationToken, Task> callback)
{
    this.OnMessageAsync(callback, new OnMessageOptions() {
        ReceiveTimeOut = this.OperationTimeout,
        AutoComplete = this.ReceiveMode == ReceiveMode.ReceiveAndDelete
    });
}
```

@jtaubensee, if you think this has some meaning, I can make a PR of it.

**Comment by jtaubensee** [comment]
@ncarandini - Did you see my previous message? I understand what you are saying, however there seems to be a disconnect.

Today, there are two different receive modes in Service Bus. Below is a description of each:
* [`PeekLock`:](https://docs.microsoft.com/rest/api/servicebus/peek-lock-message-non-destructive-read)
> This operation atomically retrieves and locks a message from a queue or subscription for processing. The message is guaranteed not to be delivered to other receivers (on the same queue or subscription only) during the lock duration specified in the queue/subscription description. When the lock expires, the message becomes available to other receivers. In order to complete processing of the message, the receiver should issue a delete command with the lock ID received from this operation. To abandon processing of the message and unlock it for other receivers, an Unlock Message command should be issued, otherwise the lock duration period can expire.

> This operation should be used in applications that require At-Least-Once delivery assurances. If the receiver does not delete the message before processing succeeds, this ensures that another receiver is able to attempt processing after the lock duration period expires.
* [`ReceiveAndDelete`:](https://docs.microsoft.com/rest/api/servicebus/receive-and-delete-message-destructive-read)
> This operation receives a message from a queue or subscription, and removes the message from that queue or subscription in one atomic operation.

With that said, I believe this is different from what you are talking about. Aren't you basically suggesting that `OnMessageOptions.AutoComplete` should be set to `false` by default? Here is a description of [`AutoComplete`](https://docs.microsoft.com/dotnet/api/microsoft.servicebus.messaging.onmessageoptions?redirectedfrom=MSDN#Microsoft_ServiceBus_Messaging_OnMessageOptions_AutoComplete).

The reason that it is set to `true`, is so that you don't always have to call `Message.Complete` inside of `OnMessage`. In the past we had some customers who forgot to do so, and then a message would keep appearing, until it was eventually dead-lettered. We really are promoting `OnMessage` as the easiest way to get up and running with Service Bus. It can be hard to develop your own receive pump, and we see customers who do it incorrectly all the time. They can very easily get a high bill because they call too often or inefficiently.

The only place that `AutoComplete` is used is [here](https://github.com/Azure/azure-service-bus-dotnet/blob/dev/src/Microsoft.Azure.ServiceBus/MessageReceivePump.cs#L208):
```csharp
if (this.messageReceiver.ReceiveMode == ReceiveMode.PeekLock &&
	this.registerHandlerOptions.AutoComplete)
{
	await this.messageReceiver.CompleteAsync(new[] { message.LockToken }).ConfigureAwait(false);
}
```

which in turn is then called [here](https://github.com/Azure/azure-service-bus-dotnet/blob/dev/src/Microsoft.Azure.ServiceBus/MessageReceivePump.cs#L173), after OnMessage has completed.

**Comment by ncarandini** [comment]
Hi @jtaubensee,
thanks for your kind explanation, now I can better understand the historical reasons of the default value of true of the AutoComplete option.

Nevertheless, while I can understand this is a safety rail for beginners, I still think this is misleading for most developers. On my opinion, using the library message pump vs. manually requesting a message shouldn't change the behaviour set by the ReceiveMode, unless I (as a developer) opt to the safety rail of AutoComplete = true.

But I know that this is not your opinion, otherwise we were not debating this here. So I want to try with a more radical approach:

Can you show me a difference in the message pump from having (ReceiveMode = PeekLook & AutoComplete = true) versus (ReceiveMode = ReceiveAndDelete) ?

If there is non difference, the net result is that AutoComplete is only a way to force the message pump (if I don't explicitly set AutoComplete option to false on the OnMessage method calling) to use the ReceiveAndDelete mode, ignoring de facto my ReceiveMode setting, only because a large number of developers do not understand the correct usage of the ReceiveMode setting.

This, to me, is kinda like the [useless box](https://www.youtube.com/watch?v=XCEtl9CfbAU) :-)

To recap:
1. Please remove AutoComplete altogether or at least minimize his impact setting it to false by default.
2. If for any reason you still think you're unwilling to remove or change the AutoComplete default value, please state this very clearly on the documentation.
3. Please do not take into account the "Inherit the value of AutoComplete from the Mode ... " section of my previous comment, because is useless after your explanation of the inner use of AutoComplete.

Many many thanks for the time you spent on this and for your reply that helped me on better understanding the AutoComplete usage and the inner working of this library.

**Comment by jtaubensee** [comment]
@ncarandini - To be clear, `AutoComplete` is only ever used for `OnMessage` / Receive pump (which are two words for the same thing). `AutoComplete` is not a property for `ReceiveAsync()`.

What we are talking about here are two different things. Just because `AutoComplete` is set to true, does not mean that the `ReceiveMode` becomes `ReceiveDelete`. It simply means that we call `Message.Complete()` for you when your lambda is done executing. If there is an error inside of that lamda, we do not call complete, and the message is available for processing again on the queue.

If you have `ReceiveMode` set `ReceiveDelete` then there is never a reason to call `Message.Complete` as the message is already removed from the queue.

Can you share your code that shows `AutoComplete` when there is an exception in your OnMessageHandler? FYI - the implementations of that message pump / OnMessage are the same between this library and the .NET Framework NuGet package.

**Comment by ncarandini** [comment]
Yes, AutoComplete is indeed a property of OnMessageOptions and it solely purpose is to call message.Complete inside the OnMessage code, only if AutoComplete is true AND the handled method works fine (i.e. no ErrorException returned). This is (finally) clear to me.

But I tested it again and again and got the same wrong result: the message is removed from the queue if AutoComplete is true AND the handled method fails.

You can easily replicate my test with a simple console app that use the actually in production NuGet package (because you say that the code is the same on this repro, we have the same bug here).

1. Create a ConsoleApp (.NET Framework)
2. Add NuGet Package "Windows.Azure.ServiceBus" (Latest stable 3.4.6)
3. Replace Program.cs code with the following code:

```csharp
using Microsoft.ServiceBus.Messaging;
using System;

namespace BugTest
{
    class Program
    {
        static void Main(string[] args)
        {
            var connectionString = "Endpoint=sb://quickitup.servicebus.windows.net/;SharedAccessKeyName=RootManageSharedAccessKey;SharedAccessKey=_<YourAccessKey>_";
            var queueName = "_YourQueueName_";

            Console.WriteLine("Test the behavior of OnMessage");
            Console.WriteLine("==============================");
            Console.Write("Connecting to the service bus...");

            var client = QueueClient.CreateFromConnectionString(connectionString, queueName);

            Console.WriteLine(" Connected.");
            Console.WriteLine();

            Console.WriteLine("Send 3 messages to the queue");
            Console.WriteLine("============================");
            Console.WriteLine();

            for (int i = 0; i < 3; i++)
            {
                var message = new BrokeredMessage($"Text message created at {DateTimeOffset.Now.ToString()}");
                client.Send(message);
                ConsoleLogSentMessage(message);
            }

            Console.WriteLine("Receive messages using OnMessage");
            Console.WriteLine("================================");
            Console.WriteLine($"The ReceiveMode of QueueClient is {client.Mode.ToString()}", ConsoleColor.Green);
            Console.WriteLine();

            client.OnMessage(receivedMessage =>
            {
                try
                {
                    // To check the OnMessage behaviour we force an Exception Error
                    throw new Exception("Exception thrown to test message pump behavior");
                }
                catch (Exception ex)
                {
                    Console.WriteLine($"Error processing MessageId {receivedMessage.MessageId}: {ex.Message}");
                    Console.WriteLine();
                }
            });

            Console.ReadKey();
        }

        static void ConsoleLogSentMessage(BrokeredMessage message)
        {
            Console.WriteLine($"[{DateTime.Now.TimeOfDay}] Message sent with MessageId: {message.MessageId}");
            Console.WriteLine($"                    TimeToLive: {message.TimeToLive.TotalSeconds}");
            Console.WriteLine();
        }

    }
}
```

**Comment by SeanFeldman** [comment]
The swallowed exception has tricked you 😃

**Comment by ncarandini** [comment]
I must wear the donkey Hat for at least a week! :-)
